### PR TITLE
[17.05] Try activating conda environment again if previous activation failed

### LINKED ISSUE #4031: Job failures when activating conda simultaneously on same compute node
The root problem is really not a galaxy issue, but it is very annoying and it would be nice if galaxy could handle this better.

If this happens, I'm seeing
```
/bioinfo/guests/mvandenb/gx/s/jsd/007/7321/tool_script.sh: line 2: findcluster: command not found
```

in the tools stderr and this in <job_working_directory>/working/conda_activate.log

```
An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : linux-64
          conda version : 4.2.12
       conda is private : False
      conda-env version : 4.2.12
    conda-build version : 2.0.9
         python version : 3.5.2.final.0
       requests version : 2.11.1
       root environment : /bioinfo/guests/mvandenb/gx/s/conda  (writable)
    default environment : /bioinfo/guests/mvandenb/gx/s/conda
       envs directories : /bioinfo/guests/mvandenb/gx/s/conda/envs
          package cache : /bioinfo/guests/mvandenb/gx/s/conda/pkgs
           channel URLs : https://conda.anaconda.org/bioconda/linux-64
                          https://conda.anaconda.org/bioconda/noarch
                          https://conda.anaconda.org/r/linux-64
                          https://conda.anaconda.org/r/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
                          https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
            config file : /data/users/mvandenb/.condarc
           offline mode : False



`$ /bioinfo/guests/mvandenb/gx/s/conda/bin/conda ..checkenv bash /bioinfo/guests/mvandenb/gx/s/conda/envs/__readtagger@0.3.19`




    Traceback (most recent call last):
      File "/bioinfo/guests/mvandenb/gx/s/conda/lib/python3.5/site-packages/conda/exceptions.py", line 479, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/bioinfo/guests/mvandenb/gx/s/conda/lib/python3.5/site-packages/conda/cli/main.py", line 94, in _main
        activate.main()
      File "/bioinfo/guests/mvandenb/gx/s/conda/lib/python3.5/site-packages/conda/cli/activate.py", line 148, in main
        conda.install.symlink_conda(prefix, context.root_dir, shell)
      File "/bioinfo/guests/mvandenb/gx/s/conda/lib/python3.5/site-packages/conda/install.py", line 536, in symlink_conda
        symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
      File "/bioinfo/guests/mvandenb/gx/s/conda/lib/python3.5/site-packages/conda/install.py", line 553, in symlink_conda_hlp
        symlink_fn(root_file, prefix_file)
    FileExistsError: [Errno 17] File exists: '/bioinfo/guests/mvandenb/gx/s/conda/bin/activate' -> '/bioinfo/guests/mvandenb/gx/s/conda/envs/__readtagger@0.3.19/bin/activate'
```

This probably has to do with the slow IO performance of our Isilion NFS server, but it would still be nice if galaxy could try again after a few seconds ... this is just super annoying when mapping a job over big collections.
FWIW, the tool_script.sh part looks like this:
```
#!/bin/bash
[ "$CONDA_DEFAULT_ENV" = "/bioinfo/guests/mvandenb/gx/s/conda/envs/__readtagger@0.3.19" ] || . /bioinfo/guests/mvandenb/gx/s/conda/bin/activate '/bioinfo/guests/mvandenb/gx/s/conda/envs/__readtagger@0.3.19' > conda_activate.log 2>&1 ; findcluster --version
```

### PR DESCRIPTION
**Pull request by mvdbeek**
That should resolve #4031.

### CODE REVIEW

**Comment by bgruening** [comment]
Good idea @mvdbeek! :+1:

File: lib/galaxy/tools/deps/resolvers/conda.py [hunk 1]
```diff
@@ -40,6 +40,19 @@
 DEFAULT_BASE_PATH_DIRECTORY = "_conda"
 DEFAULT_CONDARC_OVERRIDE = "_condarc"
 DEFAULT_ENSURE_CHANNELS = "iuc,bioconda,r,defaults,conda-forge"
+CONDA_SOURCE_CMD = """[ "$CONDA_DEFAULT_ENV" = "%s" ] ||
+MAX_TRIES=3
+COUNT=0
+while [ $COUNT -lt $MAX_TRIES ]; do
+    . %s '%s' > conda_activate.log 2>&1
+    if [ $? -eq 0 ];then
+       break
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/resolvers/conda.py]: Indentation.

File: lib/galaxy/tools/deps/resolvers/conda.py [hunk 2]
```diff
@@ -40,6 +40,19 @@
...
+    else
+        sleep 10s
+        let COUNT=COUNT+1
+    fi
+done """
```
Comment (nsoranzo) [inline lib/galaxy/tools/deps/resolvers/conda.py]: I think this may not have exit code 1 if the activation fails all 3 times.

Comment (mvdbeek) [reply lib/galaxy/tools/deps/resolvers/conda.py]: That's right, it will just continue and fail with command not found, which is OK, no ?

**Comment by jmchilton** [comment]
Thanks @mvdbeek - this seems like a good idea. I have mixed feelings on the explicit exit - it isn't something we have done in the past but it is probably better to be explicit here.

**Pull request merged**